### ISSUE TITLE
Performance decrease in query generated when using associated table in different schema

### ISSUE THREAD

**Issue by owst** [MaxLap/activerecord_where_assoc#20]
Hi, 

Thanks for the gem, it's been great for us so far. We noticed today that a less-performant query was generated when using an associated table that is in a separate schema. This seems to be known - I dug into the code/commits and found this commit that mentioned the possible performance hit: https://github.com/MaxLap/activerecord_where_assoc/commit/9b9764b791387c0cb64499274d85f495ae9b5e67

Since that original commit, have you had any other thoughts about how to handle this case more performantly? Unfortunately, I don't understand the innards of ActiveRecord well enough to precisely understand what the code is doing and why, so can't make any suggestions, but thought it was worth raising to see if it's possible to improve the generated SQL.

In an example reproduction (code below), the performance goes from 2175ms to 0.8ms if I tweak the query to avoid the redundant (for this case) subquery generated here: https://github.com/MaxLap/activerecord_where_assoc/blob/11f3e0940fe9d3f26667bd1e0c8701c692db45dd/lib/active_record_where_assoc/core_logic.rb#L371 

The reproduction setup is that we have two tables, messages and emails, and both have an (optional) polymorphic "external reference". We want to find messages that don't have an external reference (e.g. so we can create those references).

The query/plan generated by this gem as-is is:
```
> EXPLAIN ANALYZE SELECT "messages".* FROM "messages" WHERE (NOT EXISTS (SELECT 1 FROM "external"."references" WHERE "external"."references"."id" IN (SELECT "external"."references"."id" FROM "external"."references" W
                        HERE "external"."references"."resource_id" = "messages"."id" AND "external"."references"."resource_type" = 'Message' LIMIT 1)))
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| QUERY PLAN                                                                                                                                                                                      |
|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Nested Loop Anti Join  (cost=0.00..3016975.18 rows=500 width=24) (actual time=75.599..2173.598 rows=600 loops=1)                                                                                |
|   Join Filter: (SubPlan 1)                                                                                                                                                                      |
|   Rows Removed by Join Filter: 1118400                                                                                                                                                          |
|   ->  Seq Scan on messages  (cost=0.00..17.00 rows=1000 width=24) (actual time=0.010..0.272 rows=1000 loops=1)                                                                                  |
|   ->  Materialize  (cost=0.00..36.00 rows=1400 width=8) (actual time=0.073..0.127 rows=1119 loops=1000)                                                                                         |
|         ->  Seq Scan on "references"  (cost=0.00..29.00 rows=1400 width=8) (actual time=72.899..73.036 rows=1400 loops=1)                                                                       |
|   SubPlan 1                                                                                                                                                                                     |
|     ->  Limit  (cost=0.28..8.30 rows=1 width=8) (actual time=0.002..0.002 rows=0 loops=1118800)                                                                                                 |
|           ->  Index Scan using index_references_on_resource_type_and_resource_id on "references" references_1  (cost=0.28..8.30 rows=1 width=8) (actual time=0.001..0.001 rows=0 loops=1118800) |
|                 Index Cond: (((resource_type)::text = 'Message'::text) AND (resource_id = messages.id))                                                                                         |
| Planning Time: 1.624 ms                                                                                                                                                                         |
| JIT:                                                                                                                                                                                            |
|   Functions: 17                                                                                                                                                                                 |
|   Options: Inlining true, Optimization true, Expressions true, Deforming true                                                                                                                   |
|   Timing: Generation 0.983 ms, Inlining 10.318 ms, Optimization 41.902 ms, Emission 20.689 ms, Total 73.892 ms                                                                                  |
| Execution Time: 2174.822 ms                                                                                                                                                                     |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```
and if we remove the nested `IN` subquery, the much-better plan is:
```
> EXPLAIN ANALYZE SELECT "messages".* FROM "messages" WHERE (NOT EXISTS (SELECT 1 FROM "external"."references" WHERE "external"."references"."resource_id" = "messages"."id" AND "external"."references"."resource_type"
                         = 'Message' LIMIT 1))
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| QUERY PLAN                                                                                                                                                                           |
|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Hash Anti Join  (cost=24.28..50.85 rows=600 width=24) (actual time=0.546..0.747 rows=600 loops=1)                                                                                    |
|   Hash Cond: (messages.id = "references".resource_id)                                                                                                                                |
|   ->  Seq Scan on messages  (cost=0.00..17.00 rows=1000 width=24) (actual time=0.006..0.069 rows=1000 loops=1)                                                                       |
|   ->  Hash  (cost=19.28..19.28 rows=400 width=4) (actual time=0.533..0.533 rows=400 loops=1)                                                                                         |
|         Buckets: 1024  Batches: 1  Memory Usage: 23kB                                                                                                                                |
|         ->  Index Only Scan using index_references_on_resource_type_and_resource_id on "references"  (cost=0.28..19.28 rows=400 width=4) (actual time=0.445..0.492 rows=400 loops=1) |
|               Index Cond: (resource_type = 'Message'::text)                                                                                                                          |
|               Heap Fetches: 0                                                                                                                                                        |
| Planning Time: 0.444 ms                                                                                                                                                              |
| Execution Time: 0.779 ms                                                                                                                                                             |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

To reproduce, I used Postgres and used this script to create example tables:

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  gem "rails", "= 7.1.5"
  gem "activerecord_where_assoc", "= 1.2.0"

  gem "pg"
end

require "active_record/railtie"
require "minitest/autorun"

class TestApp < Rails::Application
  config.load_defaults Rails::VERSION::STRING.to_f
  config.eager_load = false
  config.logger = Logger.new($stdout)
end
Rails.application.initialize!

ActiveRecord::Schema.define do
  create_table :messages, force: true do |t|
    t.string :content
  end

  create_table :emails, force: true do |t|
    t.string :content
  end

  create_schema "external" unless schema_exists?("external")
  create_table "external.references", force: true do |t|
    t.string :resource_type
    t.integer :resource_id
    t.string :reference

    t.timestamps
  end

  add_index "external.references", [:resource_type, :resource_id], unique: true
end

class Message < ActiveRecord::Base
  has_one :external_reference, as: :resource, required: false, class_name: "External::Reference"
end

class Email < ActiveRecord::Base
  has_one :external_reference, as: :resource, required: false, class_name: "External::Reference"
end

module External
  class Reference < ActiveRecord::Base
    self.table_name = "external.references"

    belongs_to :resource, inverse_of: :external_reference, polymorphic: true
  end
end

class BugTest < ActiveSupport::TestCase
  def test_where_assoc_exists
    count = 1000

    count.times do |i|
      m = Message.create!(content: "example message")

      if (i % 10) <= 3
        m.create_external_reference!(reference: "m#{i} external ref")
      end

      e = Email.create!(content: "example email")
      e.create_external_reference!(reference: "e#{i} external ref")
    end

    messages_without_reference = Message.where_assoc_not_exists(:external_reference)

    assert_equal messages_without_reference.to_a.empty?, false
  end
end
```

I created an example DB and then ran this script with `DATABASE_URL=postgresql://user:pass@host/db_name ruby repro.rb`. 

I'm using `PostgreSQL 14.12`, with `Ruby 3.3.3`

**Comment by MaxLap** [comment]
First off, very nice detective work there!

TLDR: For you need as described here, the [ignore_limit](https://maxlap.dev/activerecord_where_assoc/ActiveRecordWhereAssoc/RelationReturningMethods.html#module-ActiveRecordWhereAssoc::RelationReturningMethods-label-3Aignore_limit+option) option should solve your problem.

```
Message.where_assoc_not_exists(:external_reference, nil, ignore_limit: true)
```

Note: If it's possible that `Message` has more than one external_reference and that you use a condition in the `where_assoc...`, then this solution will not do what you want/expect.

The limitations comes not quite from ActiveRecord, but from SQL itself and from the desire of this gem to do the right thing for `has_one`.

The short explanation is that `has_one` means take the "first" record that match the foreign key. Simple enough, your query does that just fine. However, if you where to ask is the first "external_reference" for the reference "foo" with a `where_assoc_exists(:external_reference, reference: "foo")`, you only want to test the first `external_reference` of each record for the condition, not all of the other ones that are normally skipped by `has_one`. 

To properly do that, I must create a sub-query that returns only the proper `has_one` records, and then check with those by applying your conditions. The problem is that your condition will contain the schema, such as "external.references". When I create my sub-query, it's name cannot contain a dot. So this different path has to be taken of using the `IN`.

Side note: Part of the problem comes from ActiveRecord only having `has_one` to mean to different things. A better ActiveRecord (in my opinion) would instead has `has_first` and `has_only_one`. For `has_first`, it possible/likely that there are multiple records, only take the first (basically current AR behavior). This would have the same problem as right now. But for `has_only_one`, I could assume that there are no more than one and use the same behavior as for `has_many`. Something cool is that ActiveRecord could raise errors when a `has_only_one` turns out to have more than one record, making it a real benefit for devs.

I guess I could use the faster path when there is no condition to add on top, but I feel that would be hiding the problem which can then appear later when you do need a condition... I'll need to think about it and see how complex that is to add.

Regards,

**Comment by MaxLap** [comment]
So I worked on this optimization idea which I think make sense. Can you try running your workflow on the `opti_has_ones` branch and letting me know if everything appears to work as expected?

Basically, when it's safe to do so, `has_one` will be treated as `has_many`. 

It's safe only for the last / deepest association, when no condition / block was given to the call, which matches the situation you suggested.

**Comment by owst** [comment]
Hi @MaxLap,

Thanks for the in-depth explanation. I certainly hadn't considered the point about multiple records with a condition - I can see the difficulty. 

> When I create my sub-query, it's name cannot contain a dot. 

As I say, I'm not particularly familiar with ActiveRecord internals, but I assume it's not possible to control this name?

My only other thought is, at least in my scenario, I have a unique index, ensuring there is at most one row for the `has_one` association. I'd be surprised if there was, but _is_ there enough introspection in ActiveRecord such that it could know that this constraint is enforced? If so, you could also take the`has_many` fast path there as you're guaranteed to have at most one record.

That said, thanks for pulling together the fix branch - I gave it a try, and it generates a performant query!
```
  EXPLAIN ANALYZE SELECT "messages".* FROM "messages" WHERE (NOT EXISTS (SELECT 1 FROM "external"."references" W
                        HERE "external"."references"."resource_id" = "messages"."id" AND "external"."references"."resource_type" = 'Mes
                        sage'))
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| QUERY PLAN                                                                                                                                                                           |
|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Hash Anti Join  (cost=24.28..50.85 rows=600 width=24) (actual time=0.712..1.004 rows=600 loops=1)                                                                                    |
|   Hash Cond: (messages.id = "references".resource_id)                                                                                                                                |
|   ->  Seq Scan on messages  (cost=0.00..17.00 rows=1000 width=24) (actual time=0.006..0.095 rows=1000 loops=1)                                                                       |
|   ->  Hash  (cost=19.28..19.28 rows=400 width=4) (actual time=0.655..0.655 rows=400 loops=1)                                                                                         |
|         Buckets: 1024  Batches: 1  Memory Usage: 23kB                                                                                                                                |
|         ->  Index Only Scan using index_references_on_resource_type_and_resource_id on "references"  (cost=0.28..19.28 rows=400 width=4) (actual time=0.502..0.586 rows=400 loops=1) |
|               Index Cond: (resource_type = 'Message'::text)                                                                                                                          |
|               Heap Fetches: 0                                                                                                                                                        |
| Planning Time: 3.820 ms                                                                                                                                                              |
| Execution Time: 1.069 ms                                                                                                                                                             |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

I have a few trivial comments on the PR, so I'll leave those as a review, feel free to take them or leave them.

Thanks again - if you can ping me when you've released a new version of the gem including that fix, that would be great!

**Comment by MaxLap** [comment]
The unique index check is a nice idea! It's too clever not to do it.

**Comment by MaxLap** [comment]
Version 1.2.1 has been released with both optimisations.

**Comment by owst** [comment]
Great, thank you for the super-fast response and change/release @MaxLap!
